File: 572. Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSubtree(self, s: TreeNode, t: TreeNode) -> bool:
        '''
        Recursive method, comparing with all node of s
        Time O(M*N)
        Space O(M) since we use level order traversal
        '''
        if not s and not t:
            return True
        elif not s:
            return False
        elif not t:
            return False
        queue = [s]
        while queue:
            temp = []
            for node in queue:
                res = self.check(node, t)
                if res:
                    return True
                if node.left:
                    temp.append(node.left)
                if node.right:
                    temp.append(node.right)
            queue = temp
        return False
            
            
    def check(self, node1, node2):
        if not node1 and not node2:
            return True
        elif not node1 or not node2:
            return False
        if node1.val != node2.val:
            return False
        return self.check(node1.left, node2.left) and self.check(node1.right, node2.right)

File: 572. Subtree of Another Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode s, TreeNode t) {
        return traverse(s, t);
    }
    private boolean check(TreeNode s, TreeNode t) {
        if (s == null && t==null) {
            return true; 
        } 
        if (s == null || t == null) {
            return false;
        }
        return (s.val == t.val) && check(s.left, t.left) && check(s.right, t.right);
    }
    private boolean traverse(TreeNode s, TreeNode t) {
        return s != null && (check(s,t) || traverse(s.left, t) || traverse(s.right, t));
    }
}

class Solution {
    public boolean isSubtree(TreeNode s, TreeNode t) {
        if (s == null || t == null) {
            return false;
        }
        Deque<TreeNode> dq = new LinkedList<>();
        dq.offer(s);
        while (!dq.isEmpty()) {
            TreeNode node = dq.poll();
            if (node.val == t.val) {
                if (check(node, t)) {
                    return true;
                }
            }
            if (node.left != null) {
                dq.offer(node.left);
            }
            if (node.right != null) {
                dq.offer(node.right);
            }
        }
        return false;
    }
    private boolean check(TreeNode s, TreeNode t) {
        if (s == null && t==null) {
            return true; 
        } 
        if (s == null || t == null) {
            return false;
        }
        return (s.val == t.val) && check(s.left, t.left) && check(s.right, t.right);
    }
}